### PR TITLE
ISSUE-1078 Inject event correlation information for processor runtime bolts

### LINKED ISSUE #1078: Inject event correlation information for processor runtime bolts
We're injecting information of event correlation into StreamlineEvent in test mode.
There's similar needs in runtime as well (specifically event sampling) hence we need to inject event correlation for processor runtime bolts as well.

@arunmahadevan and me discussed about approach: we decided to go with wrapping output collector for individual processor bolts (ignoring Storm stock implementation for now: for now, only PMML) since it will be applied to both runtime and test mode as well. We don't need to wrap bolts for runtime, though we still need to wrap spouts and bolts for test mode.

Source and sink doesn't need to be addressed because correlation information is not needed for them.

### PR DESCRIPTION
**Pull request by HeartSaVioR**
* change processor bolts to wrap output collector with event correlation aware output collector
* reduce chain of collector from TestRun(Window)ProcessorBolt since one of them is put to underlying bolt
* introduce EventCorrelatingWindowedOutputCollector to restrict emitting without anchor instead of hacking around
* fix broken tests due to the change

No UI change is necessary.

This closes #1078

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
Cropped result of the event sampling from this patch + ISSUE-1030 (#1031) :

```
Wed Dec 13 11:25:42 UTC 2017,31-SOURCE,1,{topic-partition=yelp_review_avro-0, offset=4456012, numFails=0},[StreamlineEvent{"dataSourceId":"31","fieldsAndValues":{"user_id":"PUFPaY9KxDAcGqfsorJp3Q","review_id":"Ya85v4eqdd6k9Od8HbQjyA","stars":4,"date":"2012-08-01","business_id":"5UmKMjUEUNdYWqANhGckJw","type":"review","votes":{"funny":0,"useful":0,"cool":0}},"auxiliaryFieldsAndValues":{},"header":{},"id":"30526c7d-da3f-43bd-8ef6-cf0972b5975c","sourceStream":"default"}]
Wed Dec 13 11:25:42 UTC 2017,31-SOURCE,1,{topic-partition=yelp_review_avro-0, offset=4456013, numFails=0},[StreamlineEvent{"dataSourceId":"31","fieldsAndValues":{"user_id":"Iu6AxdBYGR4A0wspR9BYHA","review_id":"KPvLNJ21_4wbYNctrOwWdQ","stars":5,"date":"2014-02-13","business_id":"5UmKMjUEUNdYWqANhGckJw","type":"review","votes":{"funny":0,"useful":0,"cool":0}},"auxiliaryFieldsAndValues":{},"header":{},"id":"0011086a-ff26-4d2e-9179-72776556d540","sourceStream":"default"}]
Wed Dec 13 11:25:42 UTC 2017,31-SOURCE,1,{topic-partition=yelp_review_avro-0, offset=4456014, numFails=0},[StreamlineEvent{"dataSourceId":"31","fieldsAndValues":{"user_id":"auESFwWvW42h6alXgFxAXQ","review_id":"fFSoGV46Yxuwbr3fHNuZig","stars":5,"date":"2015-10-31","business_id":"5UmKMjUEUNdYWqANhGckJw","type":"review","votes":{"funny":1,"useful":1,"cool":0}},"auxiliaryFieldsAndValues":{},"header":{},"id":"4e5ec881-fcbd-4ac4-b889-f6e3938e2e3f","sourceStream":"default"}]
Wed Dec 13 11:25:42 UTC 2017,31-SOURCE,1,{topic-partition=yelp_review_avro-0, offset=4456015, numFails=0},[StreamlineEvent{"dataSourceId":"31","fieldsAndValues":{"user_id":"qiczib2fO_1VBG8IoCGvVg","review_id":"pVMIt0a_QsKtuDfWVfSk2A","stars":3,"date":"2015-12-26","business_id":"5UmKMjUEUNdYWqANhGckJw","type":"review","votes":{"funny":0,"useful":0,"cool":0}},"auxiliaryFieldsAndValues":{},"header":{},"id":"f17e4e5a-c79d-4ef7-9f9f-f7cdf8e94e1d","sourceStream":"default"}]
Wed Dec 13 11:25:42 UTC 2017,41-RULE,2,,[StreamlineEvent{"dataSourceId":"31","fieldsAndValues":{"user_id":"PUFPaY9KxDAcGqfsorJp3Q","review_id":"Ya85v4eqdd6k9Od8HbQjyA","stars":4,"date":"2012-08-01","business_id":"5UmKMjUEUNdYWqANhGckJw","type":"review","votes":{"funny":0,"useful":0,"cool":0}},"auxiliaryFieldsAndValues":{},"header":{"rootIds":["30526c7d-da3f-43bd-8ef6-cf0972b5975c"],"parentIds":["30526c7d-da3f-43bd-8ef6-cf0972b5975c"],"sourceComponentName":"RULE"},"id":"4bea9dea-4dca-4d0f-b0bb-9c4cb71634ec","sourceStream":"kafka_stream_1"}]
Wed Dec 13 11:25:42 UTC 2017,31-SOURCE,1,{topic-partition=yelp_review_avro-0, offset=4456018, numFails=0},[StreamlineEvent{"dataSourceId":"31","fieldsAndValues":{"user_id":"bWrodc7hN_T4q2r-rolrOA","review_id":"3saY_LVFUpkAfd9t86VNdw","stars":4,"date":"2016-05-11","business_id":"5UmKMjUEUNdYWqANhGckJw","type":"review","votes":{"funny":0,"useful":0,"cool":0}},"auxiliaryFieldsAndValues":{},"header":{},"id":"b940f130-b549-4b9b-aa80-34de49916ac1","sourceStream":"default"}]
Wed Dec 13 11:25:42 UTC 2017,41-RULE,2,,[StreamlineEvent{"dataSourceId":"31","fieldsAndValues":{"user_id":"auESFwWvW42h6alXgFxAXQ","review_id":"fFSoGV46Yxuwbr3fHNuZig","stars":5,"date":"2015-10-31","business_id":"5UmKMjUEUNdYWqANhGckJw","type":"review","votes":{"funny":1,"useful":1,"cool":0}},"auxiliaryFieldsAndValues":{},"header":{"rootIds":["4e5ec881-fcbd-4ac4-b889-f6e3938e2e3f"],"parentIds":["4e5ec881-fcbd-4ac4-b889-f6e3938e2e3f"],"sourceComponentName":"RULE"},"id":"b2c30b3b-fab0-424a-9dc4-59121c2f4c12","sourceStream":"kafka_stream_1"}]
Wed Dec 13 11:25:42 UTC 2017,41-RULE,2,,[StreamlineEvent{"dataSourceId":"31","fieldsAndValues":{"user_id":"qiczib2fO_1VBG8IoCGvVg","review_id":"pVMIt0a_QsKtuDfWVfSk2A","stars":3,"date":"2015-12-26","business_id":"5UmKMjUEUNdYWqANhGckJw","type":"review","votes":{"funny":0,"useful":0,"cool":0}},"auxiliaryFieldsAndValues":{},"header":{"rootIds":["f17e4e5a-c79d-4ef7-9f9f-f7cdf8e94e1d"],"parentIds":["f17e4e5a-c79d-4ef7-9f9f-f7cdf8e94e1d"],"sourceComponentName":"RULE"},"id":"cbb6d733-58fd-4a94-969b-9fe92a528e6c","sourceStream":"kafka_stream_1"}]
Wed Dec 13 11:25:42 UTC 2017,31-SOURCE,1,{topic-partition=yelp_review_avro-0, offset=4456020, numFails=0},[StreamlineEvent{"dataSourceId":"31","fieldsAndValues":{"user_id":"I_47G-R2_egp7ME5u_ltew","review_id":"0Lua2-PbqEQMjD9r89-asw","stars":3,"date":"2014-03-29","business_id":"UsFtqoBl7naz8AVUBZMjQQ","type":"review","votes":{"funny":0,"useful":0,"cool":0}},"auxiliaryFieldsAndValues":{},"header":{},"id":"5b22e72a-e74e-4de3-9187-d7a7680cd0f5","sourceStream":"default"}]
Wed Dec 13 11:25:42 UTC 2017,31-SOURCE,1,{topic-partition=yelp_review_avro-0, offset=4456026, numFails=0},[StreamlineEvent{"dataSourceId":"31","fieldsAndValues":{"user_id":"KZuaJtFindQM9x2ZoMBxcQ","review_id":"WExNE-f93SL4D1q8s9QWKg","stars":1,"date":"2013-07-14","business_id":"cE27W9VPgO88Qxe4ol6y_g","type":"review","votes":{"funny":0,"useful":0,"cool":0}},"auxiliaryFieldsAndValues":{},"header":{},"id":"6aa706d8-3bd2-44dc-806f-9a10882aa803","sourceStream":"default"}]
```

For now I only ran with a simple topology - Kafka -> Rule -> Kafka: haven't had complicated topologies for manual tests. Will try to craft of get some afterwards.

**Comment by shahsank3t** [comment]
@arunmahadevan @HeartSaVioR 

```
Wed Dec 13 11:25:42 UTC 2017,31-SOURCE,1,{topic-partition=yelp_review_avro-0, offset=4456012, numFails=0},[StreamlineEvent{"dataSourceId":"31","fieldsAndValues":{"user_id":"PUFPaY9KxDAcGqfsorJp3Q","review_id":"Ya85v4eqdd6k9Od8HbQjyA","stars":4,"date":"2012-08-01","business_id":"5UmKMjUEUNdYWqANhGckJw","type":"review","votes":{"funny":0,"useful":0,"cool":0}},"auxiliaryFieldsAndValues":{},"header":{},"id":"30526c7d-da3f-43bd-8ef6-cf0972b5975c","sourceStream":"default"}]
```

With respect to showing the above record(s) as for how we currently show "Log Search" UI, will we be only showing this record into the table or will we also have other details(including fields for pagination), maybe like:

```
[{
  "component": "..",
  "time":"..",
  "sample": "<sample record>"
},{..}]
```

**Comment by arunmahadevan** [comment]
@shahsank3t , we will have something like  
```
{
       "timestamp": ...,
       "componentName": ...,
       "sample": ...
}
```

Will let you know exact format after we figure out the grok pattern and setup logsearch to index this.

File: streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/event/correlation/EventCorrelatingOutputCollector.java [hunk 1]
```diff
@@ -14,9 +14,10 @@
  * limitations under the License.
  **/
 
-package com.hortonworks.streamline.streams.runtime.storm.testing;
+package com.hortonworks.streamline.streams.runtime.storm.event.correlation;
 
 import com.hortonworks.streamline.streams.StreamlineEvent;
+import com.hortonworks.streamline.streams.runtime.storm.testing.StormEventCorrelationInjector;
```
Comment (arunmahadevan) [inline streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/event/correlation/EventCorrelatingOutputCollector.java]: nit: should this be moved to storm.event.correlation package ?

Comment (HeartSaVioR) [reply streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/event/correlation/EventCorrelatingOutputCollector.java]: Nice finding. Addressed.

**Comment by HeartSaVioR** [comment]
@arunmahadevan 
Just realized that the date format is not specified, which means we can't determine how to parse. It depends on the locale.

https://github.com/apache/storm/blob/b246d37419f89a33cc9895129d42952a82f667f3/storm-client/src/jvm/org/apache/storm/metric/IEventLogger.java#L48-L56

Using ',' as a delimiter is also making message hard to parse. I can't find ideal delimiter which doesn't make conflict with others, but '\t' might be a good start.

More critical thing is that we can't provide IEventLogger implementation even the intention looks like it should be pluggable. See here:

https://github.com/apache/storm/blob/b246d37419f89a33cc9895129d42952a82f667f3/storm-client/src/jvm/org/apache/storm/metric/EventLoggerBolt.java#L44-L49

It doesn't read any configuration and just instantiate FileBasedEventLogger, which makes IEventLogger implementation unusable.

**Comment by arunmahadevan** [comment]
@HeartSaVioR , Right now the implementation is fixed. If we want to provide a different one we need to patch storm. Let me know if we should change the storm implementation to allow pluggable event loggers? Other option is to parse the date and component name based on ","  and the event based on the prefix string (StreamlineEvent).

**Comment by arunmahadevan** [comment]
@HeartSaVioR , thanks for the patch merged to master. We can continue the discussion of sampling in PR #1031
